### ISSUE TITLE
pulumi refresh shows confusing changes

### ISSUE THREAD

**Issue by clstokes** [pulumi/pulumi#6243]
If a Pulumi application provisions some basic resources, the next `refresh` will likely indicate changes in the preview, but the actual refresh will not show _all_ the updates from the preview, yet the state is actually updated. This is a) confusing and b) reduces confidence in the changes that are occurring (or not).

## Expected Behavior

A `refresh` immediately after an `up` without code changes or actual resource changes should not show any changes in the refresh preview or actual refresh.

## Current Behavior

Doing an `up` then immediately doing a `refresh` shows changes in the refresh preview that do not match what is indicated in the actual refresh (e.g. `refresh --yes`), yet the state is modified.

## Steps to Reproduce

Code, logs, and state files are at https://gist.github.com/clstokes/15390df8bfab98166fad658c38e16404.

1. `up` the code at [index.ts](https://gist.github.com/clstokes/15390df8bfab98166fad658c38e16404#file-index-ts)
2. `refresh` and select `no` to see changes at [log](https://gist.github.com/clstokes/15390df8bfab98166fad658c38e16404#file-_log-L18-L35)
    - Note: there are diffs shown for each resource, but only `aws:s3:BucketPolicy` shows an actual update 
3. `refresh --yes` and see changes similar to [log](https://gist.github.com/clstokes/15390df8bfab98166fad658c38e16404#file-_log-L36-L66)
    - Note: there are diffs shown for each resource, but only `aws:s3:BucketPolicy` shows an actual update 
4. `pre --diff` - shows a difference to `aws:s3:BucketPolicy`, but seems to be `"true" => true` in the actual JSON policy document
    - Note: This might be a different issue and not the root cause here?

Full steps and output are at https://gist.github.com/clstokes/15390df8bfab98166fad658c38e16404#file-_log.

## Context (Environment)

Versions tested are below, but per user reports this is not specific to these versions.

```
p-bucket-refresh % pulumi version
v2.19.0
p-bucket-refresh % npm ls
p-bucket-refresh@ /Users/clstokes/cc/scratch/p-bucket-refresh
├── @pulumi/aws@3.26.1
├── @pulumi/awsx@0.22.0
├── @pulumi/pulumi@2.19.0
└── @types/node@10.17.51
```

**Comment by clstokes** [comment]
Related to https://github.com/pulumi/pulumi/issues/6241 ?

**Comment by lukehoban** [comment]
A simple concrete example here.

If I deploy a single S3 bucket, I get this state file after initial `pulumi up`.

```json
{
    "version": 3,
    "deployment": {
        "manifest": {
            "time": "2021-02-25T06:02:15.042117+11:00",
            "magic": "a2feae0b7db5df9aa0989f710df059816f405615e1910235d53aa610ad01bdd3",
            "version": "v2.21.2"
        },
        "secrets_providers": {
            "type": "service",
            "state": {
                "url": "https://api.pulumi.com",
                "owner": "lukehoban",
                "project": "s3refresh",
                "stack": "dev"
            }
        },
        "resources": [
            {
                "urn": "urn:pulumi:dev::s3refresh::pulumi:pulumi:Stack::s3refresh-dev",
                "custom": false,
                "type": "pulumi:pulumi:Stack",
                "outputs": {
                    "bucketName": "my-bucket-790470e"
                }
            },
            {
                "urn": "urn:pulumi:dev::s3refresh::pulumi:providers:aws::default_3_30_1",
                "custom": true,
                "id": "1ef9ee77-7490-410b-8b3f-9fd7cdd7d446",
                "type": "pulumi:providers:aws",
                "inputs": {
                    "region": "us-east-1",
                    "version": "3.30.1"
                },
                "outputs": {
                    "region": "us-east-1",
                    "version": "3.30.1"
                }
            },
            {
                "urn": "urn:pulumi:dev::s3refresh::aws:s3/bucket:Bucket::my-bucket",
                "custom": true,
                "id": "my-bucket-790470e",
                "type": "aws:s3/bucket:Bucket",
                "inputs": {
                    "__defaults": [
                        "acl",
                        "bucket",
                        "forceDestroy"
                    ],
                    "acl": "private",
                    "bucket": "my-bucket-790470e",
                    "forceDestroy": false
                },
                "outputs": {
                    "accelerationStatus": "",
                    "acl": "private",
                    "arn": "arn:aws:s3:::my-bucket-790470e",
                    "bucket": "my-bucket-790470e",
                    "bucketDomainName": "my-bucket-790470e.s3.amazonaws.com",
                    "bucketRegionalDomainName": "my-bucket-790470e.s3.amazonaws.com",
                    "corsRules": [],
                    "forceDestroy": false,
                    "grants": [],
                    "hostedZoneId": "Z3AQBSTGFYJSTF",
                    "id": "my-bucket-790470e",
                    "lifecycleRules": [],
                    "loggings": [],
                    "objectLockConfiguration": null,
                    "region": "us-east-1",
                    "replicationConfiguration": null,
                    "requestPayer": "BucketOwner",
                    "serverSideEncryptionConfiguration": null,
                    "tags": {},
                    "versioning": {
                        "enabled": false,
                        "mfaDelete": false
                    },
                    "website": null
                },
                "parent": "urn:pulumi:dev::s3refresh::pulumi:pulumi:Stack::s3refresh-dev",
                "provider": "urn:pulumi:dev::s3refresh::pulumi:providers:aws::default_3_30_1::1ef9ee77-7490-410b-8b3f-9fd7cdd7d446"
            }
        ]
    }
}
```

If I then refresh, I see:

```
Previewing refresh (dev)

View Live: https://app.pulumi.com/lukehoban/s3refresh/dev/previews/25076388-bbb1-4ac4-95dd-7df3b0e4f00d

     Type                 Name           Plan     Info
     pulumi:pulumi:Stack  s3refresh-dev
     └─ aws:s3:Bucket     my-bucket               [diff: +accelerationStatus,arn,corsRules,grants,hostedZoneId,lifecycleRules,loggings,requestPayer

Resources:
    2 unchanged
```

And after doing that refresh, the state file is:

```json
{
    "version": 3,
    "deployment": {
        "manifest": {
            "time": "2021-02-25T06:03:10.388028+11:00",
            "magic": "a2feae0b7db5df9aa0989f710df059816f405615e1910235d53aa610ad01bdd3",
            "version": "v2.21.2"
        },
        "secrets_providers": {
            "type": "service",
            "state": {
                "url": "https://api.pulumi.com",
                "owner": "lukehoban",
                "project": "s3refresh",
                "stack": "dev"
            }
        },
        "resources": [
            {
                "urn": "urn:pulumi:dev::s3refresh::pulumi:pulumi:Stack::s3refresh-dev",
                "custom": false,
                "type": "pulumi:pulumi:Stack",
                "outputs": {
                    "bucketName": "my-bucket-790470e"
                }
            },
            {
                "urn": "urn:pulumi:dev::s3refresh::pulumi:providers:aws::default_3_30_1",
                "custom": true,
                "id": "1ef9ee77-7490-410b-8b3f-9fd7cdd7d446",
                "type": "pulumi:providers:aws",
                "inputs": {
                    "region": "us-east-1",
                    "version": "3.30.1"
                },
                "outputs": {
                    "region": "us-east-1",
                    "version": "3.30.1"
                }
            },
            {
                "urn": "urn:pulumi:dev::s3refresh::aws:s3/bucket:Bucket::my-bucket",
                "custom": true,
                "id": "my-bucket-790470e",
                "type": "aws:s3/bucket:Bucket",
                "inputs": {
                    "__defaults": [
                        "acl",
                        "bucket",
                        "forceDestroy"
                    ],
                    "accelerationStatus": "",
                    "acl": "private",
                    "arn": "arn:aws:s3:::my-bucket-790470e",
                    "bucket": "my-bucket-790470e",
                    "corsRules": [],
                    "forceDestroy": false,
                    "grants": [],
                    "hostedZoneId": "Z3AQBSTGFYJSTF",
                    "lifecycleRules": [],
                    "loggings": [],
                    "objectLockConfiguration": null,
                    "replicationConfiguration": null,
                    "requestPayer": "BucketOwner",
                    "serverSideEncryptionConfiguration": null,
                    "tags": {},
                    "versioning": {
                        "enabled": false,
                        "mfaDelete": false
                    },
                    "website": null
                },
                "outputs": {
                    "accelerationStatus": "",
                    "acl": "private",
                    "arn": "arn:aws:s3:::my-bucket-790470e",
                    "bucket": "my-bucket-790470e",
                    "bucketDomainName": "my-bucket-790470e.s3.amazonaws.com",
                    "bucketRegionalDomainName": "my-bucket-790470e.s3.amazonaws.com",
                    "corsRules": [],
                    "forceDestroy": false,
                    "grants": [],
                    "hostedZoneId": "Z3AQBSTGFYJSTF",
                    "id": "my-bucket-790470e",
                    "lifecycleRules": [],
                    "loggings": [],
                    "objectLockConfiguration": null,
                    "region": "us-east-1",
                    "replicationConfiguration": null,
                    "requestPayer": "BucketOwner",
                    "serverSideEncryptionConfiguration": null,
                    "tags": {},
                    "versioning": {
                        "enabled": false,
                        "mfaDelete": false
                    },
                    "website": null
                },
                "parent": "urn:pulumi:dev::s3refresh::pulumi:pulumi:Stack::s3refresh-dev",
                "provider": "urn:pulumi:dev::s3refresh::pulumi:providers:aws::default_3_30_1::1ef9ee77-7490-410b-8b3f-9fd7cdd7d446"
            }
        ]
    }
}
```

Notably:
1. The `outputs` didn't change at all between the "before" and "after".
2. The `inputs` did change to add a lot of properties.  These are in two classes: (a) Optional+Computeds that are getting added to inputs and (b) Optional TypeList/TypeSet that are being newly populated into inputs with `[]`.
3. The diff displayed in the refresh preview appears to be displaying the change in the inputs - even though we intend to be displaying the diff in the outputs.

(2a) is tracked in #6241.

(2b) seems to be another issue we likely should address either as part of #6241 or this issue.

(3) seems to be an issue specific to the rendering of diffs, which might be indirectly addressed by #6241, but likely indicates some other issue with us incorrectly displaying diffs of inputs instead of outputs.

**Comment by lukehoban** [comment]
Re-ran this with the changes from https://github.com/pulumi/pulumi-terraform-bridge/pull/330 and https://github.com/pulumi/pulumi/pull/6568, and it now behaves as expected.

The first `refresh` shows:

```
     Type                    Name                   Plan       Info
     pulumi:pulumi:Stack     refresh-test-luke-dev             
 ~   ├─ aws:s3:BucketPolicy  my-bucket-policy       update     [diff: ~policy]
     ├─ aws:s3:Bucket        my-bucket                         
     └─ aws:ec2:Instance     my-instance                       
 
Resources:
    ~ 1 to update
    3 unchanged

Do you want to perform this refresh?
No resources will be modified as part of this refresh; just your stack's state will be. details
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::refresh-test-luke::pulumi:pulumi:Stack::refresh-test-luke-dev]
    ~ aws:s3/bucketPolicy:BucketPolicy: (update)
        [id=my-bucket-aa02311]
        [urn=urn:pulumi:dev::refresh-test-luke::aws:s3/bucketPolicy:BucketPolicy::my-bucket-policy]
        [provider=urn:pulumi:dev::refresh-test-luke::pulumi:providers:aws::default_3_33_0::b3f2c140-1b44-4efe-b73c-54ed6c7cf823]
        --outputs:--
      ~ policy: "{\"Version\":\"2012-10-17\",\"Id\":\"MYBUCKETPOLICY\",\"Statement\":[{\"Sid\":\"\",\"Effect\":\"Deny\",\"Principal\":\"*\",\"Action\":\"s3:*\",\"Resource\":\"arn:aws:s3:::my-bucket-aa02311/*\",\"Condition\":{\"Null\":{\"aws:MultiFactorAuthAge\":true}}}]}" => "{\"Version\":\"2012-10-17\",\"Id\":\"MYBUCKETPOLICY\",\"Statement\":[{\"Sid\":\"\",\"Effect\":\"Deny\",\"Principal\":\"*\",\"Action\":\"s3:*\",\"Resource\":\"arn:aws:s3:::my-bucket-aa02311/*\",\"Condition\":{\"Null\":{\"aws:MultiFactorAuthAge\":\"true\"}}}]}"
```

As noted in the original, this is due to the use of `true` in the code, while AWS normalizes to `"true"`.  To avoid a diff on refreshes, the user code must change to use `"true"` as well, and that aspect is expected.

There are no longer any other extraneous refresh diffs.
